aten/src/ATen/cpu/vec/vec256/vec256_mask.h

```diff
@@ -46,14 +101,97 @@ struct VecMaskCast<int, 1, float, 1> {
+  }
+};
+
+template <>
+struct VecMaskCast<int64_t, 2, double, 2> {
+  static inline VecMask<int64_t, 2> apply(const VecMask<double, 2>& vec_mask) {
+    VectorizedN<int64_t, 2> result;
+    result[0] = _mm256_castpd_si256(vec_mask[0]);
+    result[1] = _mm256_castpd_si256(vec_mask[1]);
+    return result;
+  }
+};
 
+template <>
+struct VecMaskCast<double, 2, int64_t, 2> {
+  static inline VecMask<double, 2> apply(const VecMask<int64_t, 2>& vec_mask) {
+    VectorizedN<double, 2> result;
+    result[0] = _mm256_castsi256_pd(vec_mask[0]);
+    result[1] = _mm256_castsi256_pd(vec_mask[1]);
+    return result;
+  }
+};
 template <typename dst_t>
 struct VecMaskCast<dst_t, 1, int64_t, 2> {
   static inline VecMask<dst_t, 1> apply(const VecMask<int64_t, 2>& vec_mask) {
     auto int_vec = convert<int, 1, int64_t, 2>(VectorizedN<int64_t, 2>(vec_mask));
     return VecMask<int, 1>(int_vec).cast<dst_t, 1>();
   }
 };
+template <typename mask_t>
+struct VecMaskCast<int64_t, 2, mask_t, 1> {
+  static inline VecMask<int64_t, 2> apply(const VecMask<mask_t, 1>& vec_mask) {
+    auto int_mask = vec_mask.template cast<int, 1>();
+    auto int64_vec = convert<int64_t, 2, int, 1>(VectorizedN<int, 1>(int_mask[0]));
+    return int64_vec;
+  }
+};
+template <>
+struct VecMaskCast<double, 2, float, 1> {
+  static inline VecMask<double, 2> apply(const VecMask<float, 1>& vec_mask) {
+    auto int64_mask = VecMaskCast<int64_t, 2, float, 1>::apply(vec_mask);
+    return VecMaskCast<double, 2, int64_t, 2>::apply(int64_mask);
+  }
+};
+template <>
+struct VecMaskCast<float, 1, double, 2> {
+  static inline VecMask<float, 1> apply(const VecMask<double, 2>& vec_mask) {
+    auto int64_mask = VecMaskCast<int64_t, 2, double, 2>::apply(vec_mask);
+    return VecMaskCast<float, 1, int64_t, 2>::apply(int64_mask);
+  }
+};
+
+template <typename mask_t>
+struct VecMaskCast<int64_t, 4, mask_t, 2> {
+  static inline VecMask<int64_t, 4> apply(const VecMask<mask_t, 2>& vec_mask) {
+    auto result = at::vec::VectorizedN<int64_t, 4>();
+    auto int_mask = vec_mask.template cast<int, 2>();
+    auto int64_vec = convert<int64_t, 2, int, 1>(VectorizedN<int, 1>(int_mask[0]));
+    result[0] = int64_vec[0];
+    result[1] = int64_vec[1];
+    int64_vec = convert<int64_t, 2, int, 1>(VectorizedN<int, 1>(int_mask[1]));
+    result[2] = int64_vec[0];
+    result[3] = int64_vec[1];
+    return VecMask<int64_t, 4>(result);
+  }
+};
+
+template <typename dst_t>
+struct VecMaskCast<dst_t, 2, int64_t, 4> {
+  static inline VecMask<dst_t, 2> apply(const VecMask<int64_t, 4>& vec_mask) {
+    auto result = VecMask<int, 2>();
+    auto int64_vec = at::vec::VectorizedN<int64_t, 2>();
+    int64_vec[0] = vec_mask[0];
+    int64_vec[1] = vec_mask[1];
+    result[0] = convert<int, 1, int64_t, 2>(int64_vec);
+    int64_vec[0] = vec_mask[2];
+    int64_vec[1] = vec_mask[3];
+    result[1] = convert<int, 1, int64_t, 2>(int64_vec);
+    return VecMask<int, 2>(result).cast<dst_t, 2>();
+  }
+};
```

Comment:
ditto: we may consider to generalize it as sizes for `N` and `2*N` and do for loop calling specialized sizes of `1` and `2` here.

Reply:
Revised according to suggestions